source/souper.1
.\" $Id: souper.1 $
.\"
.de EX		\"Begin example
.ne 5
.if n .sp 1
.if t .sp .5
.nf
.in +5n
..
.de EE		\"End example
.fi
.in -5n
.if n .sp 1
.if t .sp .5
..
.TH Souper 1 " "
.ds LH "Souper User Manual - version 1.5
.ds CH "
.ds RH "
.ll 78n
.na
.ce 1
.SH Souper \- transfer mail and news to SOUP
.LP
Souper is a program that transfers mail and news from a POP3 mail server
and NNTP news server respectively to SOUP packets.
It can also send messages in SOUP reply packets.
It runs on OS/2 Warp with the IBM Internet Connection for OS/2.
.SH INSTALLATION
.LP
Copy the souper.exe file to a directory in your PATH.
.LP
Edit your \\tcpip\\etc\\services file.
Look for a line beginning with the word "pop3".
If the file does not contain such a line, add the line
.EX
pop3             110/tcp
.EE
to the file.
.LP
Souper is designed to work with the IBM Internet Connection for OS/2.
In your PPP/SLIP settings notebook, you must fill in the entries for
.EX
Your Host Name:
Your Domain Name:
News Server:
Mail Gateway:
POP Mail Server:
Reply Domain:
Reply (Mail) ID:
POP Login ID:
POP Password:
.EE
Ask your Internet service provider for the correct values for these settings.
.SH ENVIRONMENT VARIABLES
.LP
You can set these environment variables to change the default behaviour
of the program.
.TP 10
HOME
Specifies the user's home directory.
If not set, this defaults to the current directory.
The program looks for the newsrc file in the home directory.
.TP
NNTPSERVER
Specify the NNTP news server, overriding the news server given in the
PPP/SLIP settings notebook.
.TP
MAILER
Specify the command to execute to send mail.
If this is not set, then Souper sends the mail directly to the mail server
specified in the PPP/SLIP settings notebook.
.TP
POSTER
Specify the command to execute to post news.
If this is not set, then Souper posts the news directly to the news server
specified in the PPP/SLIP settings notebook.
.SH CREATE NEWSRC FILE
.LP
Use a text editor to create a file named
.B newsrc
in your home directory.
List the newsgroups you want to transfer, one per line, and end each
line with a colon.
For example:
.EX
comp.answers:
news.answers:
rec.humor.funny:
.EE
The program records the article numbers it has fetched by updating
this file.
.LP
If you use Yarn, be sure to give Souper a separate
.B newsrc
file from the one used by Yarn.
The files must be separate and are used for different purposes.
The Souper
.B newsrc
file records the articles you have downloaded from the news server.
The Yarn
.B newsrc
file records the articles you have read with the Yarn news reader.
.SH CREATE KILL FILE
.LP
A kill file specifies criteria used to kill news articles,
that is, exclude them from the SOUP packet.
You can kill articles that have a specific subject, are from
a specific poster, or contain a particular string anywhere in the header.
.LP
By default, the program looks for a file named kill in your home directory.
An entry in the kill file has the format:
.EX
<newsgroup> {
  <where> <pattern>
  ...
}
.EE
where
.TP 10
<newsgroup>
is the name of the newsgroup in which the search patterns apply.
If the name is "all", the search patterns apply in all newsgroups.
.TP
<where>
specifies where to search in the article.
"from" searches the From line.
"subject" searches the Subject line.
"header" searches all lines in the article header.
.TP
<pattern>
is the string to search for.
All searches are case-insensitive.
.ne 9
.LP
Here is an example kill file:
.EX
all {
  From netkook
}
comp.os.os2.advocacy {
  Subject Microsoft
  Subject crash OS/2
}
.EE
In all newsgroups, articles from netkook are killed.
Also in the comp.os.os2.advocacy newsgroup, articles with a Subject
containing the string "Microsoft" or "crash OS/2" are killed.
.SH OPERATION
.LP
To fetch your mail and news, run the command
.EX
souper
.EE
You can specify a different POP3 mail server than the one specified in
the PPP/SLIP settings notebook by instead running the command
.EX
souper <mailhost> <userid> <password>
.EE
where
.TP 10
<mailhost>
is the name of the host running the POP3 mail server
.TP
<userid>
is your POP3 user ID
.TP
<password>
is your POP3 password
.LP
If you want to transfer only news, specify the -m option to disable the
mail transfer.
.EX
souper -m
.EE
If you want to transfer only mail, specify the -n option to disable the
news transfer.
.EX
souper -n
.EE
.LP
The program creates the AREAS and *.MSG files of the SOUP packet in
the current directory.
Here's a typical batch file designed to import news into Yarn:
.EX
cd \\temp
souper
import -u
.EE
And to export messages from Yarn:
.EX
cd \\temp
unzip -o \\home\\reply.zip
del \\home\\reply.zip
souper -s
.EE
.SH OPTIONS
.LP
The program accepts the following command line options.
.TP 10
\-a
Check for new newsgroups.
If any newgroups were created since the last time
.B Souper
was run, add the new newsgroups to your newsrc file.
.TP 10
\-c \fIn\fR
Catch up news.
Mark every article in every newsgroup as read except for the last
.I n
articles.
.TP
\-h \fIdir\fR
Set home directory.
This option overrides any HOME environment setting.
If the HOME environment variable is not set,
the default is the current directory.
.TP
\-i
Do not read the configuration from the IBM Internet Connection for OS/2
settings notebook.
If the use this option, you must specify the required mail and news
host information through the command line and environment variables.
.TP
\-k \fIn\fR
Set maximum news packet size in Kbytes.
The default is 2048 Kbytes.
If you set this to 0, the packet size is unlimited.
.TP
\-K \fIfile\fR
Specify another kill file.
The default is the kill file in the home directory.
.TP
\-l \fIn\fR
Kill articles that contain more than
.I n
lines in the body. 
.TP
\-m
Do not get mail.
.TP
\-n
Do not get news.
.TP
\-N \fIfile\fR
Specify another newsrc file.
The default is the newsrc file in the home directory.
.TP
\-r
Set read only mode.
Do not empty your POP3 mailbox or update the newsrc file.
.TP
\-s
Send replies.
The program expects the SOUP reply files in the current directory.
.TP
\-u
When fetching news, create SOUP summaries instead of transfering
entire articles.
.TP
\-x
Do not process news Xref headers.
.SH ACKNOWLEDGEMENTS
.LP
This program is derived from popclient by Carl Harris <ceharris@vt.edu>
and uqwk by Steve Belczyk <seb3@gte.com>.
.ne 8
.SH STANDARD DISCLAIMER
.LP
The author makes no representations about the accuracy or suitability of
this material for any purpose.
It is provided "as is", without any express or implied warranties.
The author will assume no liability for damages either from the direct use
of this product or as a consequence of the use of this product.



CHANGES.txt
VERSION 1.5

- The program is now compiled with Watcom C++ 10.0b.  It no longer requires
  the emx run-time DLLs.

- If an outgoing message has a Resent-To header, the program now sends the
  message to the address on that header instead of the usual To, Cc and Bcc
  headers.

- The download summary mode now marks articles as read.

- Fix: When sending SMTP mail, the program included comments in the address
  field of MAIL FROM and RCPT TO commands.

VERSION 1.4

- Now look up the service names "pop3" and then "pop" in that order in the
  services file.

- For killed articles, now display the header matching the kill expression.

- Fix: The program tried to establish an SMTP connection to the host
  specified by the POP Mail Server entry in the settings notebook.  It
  now makes an SMTP connection to the host specified in the Mail Gateway
  setting.

VERSION 1.3

- Now reads the configuration from the IBM Internet Connection for OS/2
  settings notebook.  Use the -i command line option to ignore the settings.

- The program can now send messages using SMTP and post articles using NNTP.

- Added -a option to check for new newsgroups.

- Added -l option to kill articles longer than a specified number of lines.

- Added -u option to create SOUP news summary.

- If a COMMANDS file exists in the current directory, the program now
  processes SOUP sendme commands in the file.

VERSION 1.2

- Now display percentage progress indicator during download.

- Fix: The first article was skipped if the user newly subscribed to a
  newsgroup in which the lowest article number is 1.

VERSION 1.1

- Added simple kill file support.


source/souper.c
/* $Id: souper.c 1.6 1996/05/18 21:13:54 cthuang Exp $
 *
 * Fetch mail and news using POP3 and NNTP into a SOUP packet.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>
#include "souper.h"

#ifdef __WIN32__
#include <fcntl.h>
#include <winsock.h>
WSADATA WSAData;
#endif

/* getopt declarations */
#if HAVE_GETOPT_H
#include <getopt.h>
#else
extern int getopt(int argc, char *const *argv, const char *shortopts);
extern char *optarg;
extern int optind;
#endif

/* global data */
char progname[] = "souper";

#ifdef __OS2__
char doIni = 1;		/* if TRUE, read TCPOS2.INI file */
#endif
char *nntpServer;
char *mailGateway;

/* program options */
enum Mode { RECEIVE, SEND, CATCHUP } mode = RECEIVE;
char doMail = 1;
char doNews = 1;
char doXref = 1;
char doSummary = 0;
char doNewGroups = 0;
char readOnly = 0;
long maxBytes = 2048*1024L;
char *homeDir = NULL;
char newsrcFile[FILENAME_MAX];
char killFile[FILENAME_MAX];
int maxLines = 0;
int catchupCount;
char *popServer, *popUser, *popPassword;

/* Try to allocate some memory.  Exit if unsuccessful.
 */
void *
xmalloc (size_t sz)
{
    void *p;

    if ((p = malloc(sz)) == NULL) {
	fputs("out of memory\n", stderr);
	exit(EXIT_FAILURE);
    }
    return p;
}

/* Copy string to allocated memory.  Exit if unsuccessful.
 */
char *
xstrdup (const char *s)
{
    return strcpy((char *)xmalloc(strlen(s)+1), s);
}

static void
setDefaults (void)
{
    char *s;

    if (!homeDir)
	if ((homeDir = getenv("HOME")) == NULL)
	    homeDir = ".";

    if ((s = getenv("NNTPSERVER")) != NULL)
	nntpServer = s;

#ifdef __OS2__
    sprintf(newsrcFile, "%s/newsrc", homeDir);
    sprintf(killFile, "%s/kill", homeDir);
#else
    sprintf(newsrcFile, "%s/.newsrc", homeDir);
    sprintf(killFile, "%s/.kill", homeDir);
#endif
}

static void
usage (void)
{
    fputs("Souper v1.5 - transfer POP3 mail and NNTP news to SOUP\n", stderr);
    fprintf(stderr, "usage: %s [options] [mailhost userid password]\n", progname);
    fputs("  -a       Add new newsgroups to newsrc file\n", stderr);
    fputs("  -c n     Mark every article as read except for the last n in each newsgroup\n", stderr);
    fputs("  -h dir   Set home directory\n", stderr);
#ifdef __OS2__
    fputs("  -i       Do not read Internet Connection for OS/2 settings\n", stderr);
#endif
    fputs("  -k n     Set maximum news packet size in Kbytes\n", stderr);
    fputs("  -l n     Kill articles longer than n lines\n", stderr);
    fputs("  -m       Do not get mail\n", stderr);
    fputs("  -n       Do not get news\n", stderr);
    fputs("  -N file  Set newsrc file\n", stderr);
    fputs("  -K file  Set kill file\n", stderr);
    fputs("  -r       Read only mode.  Do not delete mail or update newsrc\n", stderr);
    fputs("  -s       Send replies\n", stderr);
    fputs("  -u       Create news summary\n", stderr);
    fputs("  -x       Do not process news Xref headers\n", stderr);
    exit(EXIT_FAILURE);
}

static void
parseOptions (int argc, char **argv)
{
    int c;

    while ((c = getopt(argc, argv, "ac:h:iK:k:l:mN:nrsux")) != EOF) {
	switch (c) {
	case 'a':
	    doNewGroups = 1;
	    break;
	case 'c':
	    mode = CATCHUP;
	    catchupCount = atoi(optarg);
	    break;
	case 'h':
	    homeDir = optarg;
	    setDefaults();
	    break;
#ifdef __OS2__
	case 'i':
	    doIni = 0;
	    break;
#endif
	case 'K':
	    strcpy(killFile, optarg);
	    break;
	case 'k':
	    maxBytes = atol(optarg) * 1024L;
	    break;
	case 'l':
	    maxLines = atoi(optarg);
	    break;
	case 'm':
	    doMail = 0;
	    break;
	case 'N':
	    strcpy(newsrcFile, optarg);
	    break;
	case 'n':
	    doNews = 0;
	    break;
	case 'r':
	    readOnly = 1;
	    break;
	case 's':
	    mode = SEND;
	    break;
	case 'u':
	    doNews = 1;
	    doSummary = 1;
	    break;
	case 'x':
	    doXref = 0;
	    break;
	default:
	    usage();
	}
    }

}

#ifdef __OS2__
#define INCL_WINSHELLDATA
#include <os2.h>

static void
readTcpIni (void)
{
    HAB hab;
    HINI hini;
    char *etc;
    char buf[BUFSIZ], curConnect[20], host[40], domain[40];

    etc = getenv("ETC");
    if (etc == NULL) {
	fputs("Must set ETC\n", stderr);
	exit(EXIT_FAILURE);
    }
    sprintf(buf, "%s\\TCPOS2.INI", etc);

    hab = WinInitialize(0);
    hini = PrfOpenProfile(hab, buf);
    if (hini == NULLHANDLE) {
	fprintf(stderr, "Cannot open profile %s\n", buf);
	exit(EXIT_FAILURE);
    }

    PrfQueryProfileString(hini, "CONNECTION", "CURRENT_CONNECTION", NULL,
                          curConnect, sizeof(curConnect));

#if 0
    PrfQueryProfileString(hini, curConnect, "HOSTNAME", NULL, host,
	sizeof(host));
    PrfQueryProfileString(hini, curConnect, "DOMAIN_NAME", NULL, domain,
	sizeof(domain));
    sprintf(buf, "%s.%s", host, domain);
    hostName = xstrdup(buf);
#endif

    PrfQueryProfileString(hini, curConnect, "POPSRVR", NULL, buf, sizeof(buf));
    popServer = xstrdup(buf);
    PrfQueryProfileString(hini, curConnect, "POP_ID", NULL, buf, sizeof(buf));
    popUser = xstrdup(buf);
    PrfQueryProfileString(hini, curConnect, "POP_PWD", NULL, buf, sizeof(buf));
    popPassword = xstrdup(buf);

    if (!nntpServer) {
	PrfQueryProfileString(hini, curConnect, "DEFAULT_NEWS", NULL, buf, sizeof(buf));
	nntpServer = xstrdup(buf);
    }

    PrfQueryProfileString(hini, curConnect, "MAIL_GW", NULL, buf, sizeof(buf));
    mailGateway = xstrdup(buf);
#if 0
    PrfQueryProfileString(hini, curConnect, "REPLY_DOMAIN", NULL, buf, sizeof(buf));
    replyDomain = xstrdup(buf);
#endif

    PrfCloseProfile(hini);
    WinTerminate(hab);
}
#endif

int
main (int argc, char **argv)
{
#ifdef __WIN32__ /* Set up to save the old signal */
   typedef void (*Sig_t)(int);
   Sig_t functionCall;
#endif
#if 0
#ifdef __OS2__
    progname = strrchr(argv[0], '\\');
#else
    progname = strrchr(argv[0], '/');
#endif
    if (progname == NULL)
	progname = argv[0];
    else
	++progname;
#endif

#ifdef __WIN32__
{
    int err;
    WORD wVerReq;

    wVerReq = MAKEWORD(1,1);
   
    err = WSAStartup(wVerReq, &WSAData);
    if (err != 0) {
        printf("Help!!!");
        return -1;
    }
}
#endif

    setDefaults();
    parseOptions(argc, argv);
#ifdef __OS2__
    if (doIni)
	readTcpIni();
#endif

    switch (mode) {
    case RECEIVE:
	if (doMail) {
	    if (argc - optind == 3) {
		popServer = argv[optind];
		popUser = argv[optind+1];
		popPassword = argv[optind+2];
	    } else {
#ifndef __OS2__
		usage();
#endif
	    }
	}

	openAreas();
#ifdef __WIN32__
	/* Save the old signal handler. */
        functionCall = signal(SIGINT, closeAreasOnSignal);
#else
	signal(SIGINT, closeAreasOnSignal);
#endif  
	if (doMail) getMail(popServer, popUser, popPassword);
	if (doNews) {
	    if (doSummary) {
		sumNews();
	    } else {
		if (!getNews())
		    exit(EXIT_FAILURE);
	    }
	}
	closeAreas();
	break;

    case SEND:
	if (argc - optind >= 1) {
	    mailGateway = argv[optind];
	}
	sendReply();
	break;

    case CATCHUP:
	catchupNews(catchupCount);
	break;
    }

#ifdef __WIN32__
    signal(SIGINT, functionCall);	/* Restore the old signal handler. */
    WSACleanup();
#endif
    return EXIT_SUCCESS;
}
